backend/wallet/services/transaction.py

```diff
@@ -124,16 +150,95 @@ def process_incoming_transaction(
+            f"Travel Rule metadata decode failed: Transaction {transaction.id} with reference ID {reference_id} "
             f"should have amount: {transaction.amount}, status: {TransactionStatus.OFF_CHAIN_READY}, "
             f"source address: {transaction.source_address}, destination address: {transaction.destination_address},"
             f" but received transaction has amount: {amount}, status: {transaction.status}, "
             f"source address: {sender_address}, destination address: {receiver_address}"
         )
 
-    if not receiver_id:
-        log_execution("Incoming transaction had no metadata. crediting inventory")
-        receiver_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id
+    if isinstance(metadata, diem_types.Metadata__RefundMetadata) and isinstance(
+        metadata.value, diem_types.RefundMetadata__RefundMetadataV0
+    ):
+        logger.info(
+            f"process_incoming_transaction refund metadata [{blockchain_version}]: "
+            f"sender: {sender_address}, receiver: {receiver_address}, "
+            f"seq: {sequence}, amount: {amount}"
+        )
+        txn_version = metadata.value.value.transaction_version
+        reason = metadata.value.value.reason
+        payment_type = TransactionType.REFUND
+
+        original_txn = get_transaction_by_blockchain_version(txn_version)
+
+        logger.info(f"Refund metadata: txn version: {txn_version}, reason: {reason}")
+
+        # Cannot find the referred original transaction
+        if original_txn is None:
+            logger.info(
+                f"Refund metadata invalid: txn version: {txn_version} does not exist"
+                f"Sending refund back to sender {sender_address}"
+            )
+            # Record the received transaction as a regular transaction to inventory account
+            inventory_account_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id
+            add_transaction(
+                amount=amount,
+                currency=currency,
+                payment_type=TransactionType.EXTERNAL,
+                status=TransactionStatus.FAILED,
+                source_address=sender_address,
+                destination_id=inventory_account_id,
+                destination_address=receiver_address,
+                sequence=sequence,
+                blockchain_version=blockchain_version,
+            )
+            return
+
+        if isinstance(reason, diem_types.RefundReason__InvalidSubaddress):
+            refund_reason = RefundReason.INVALID_SUBADDRESS
+        elif isinstance(reason, diem_types.RefundReason__UserInitiatedPartialRefund):
+            refund_reason = RefundReason.USER_INITIATED_PARTIAL_REFUND
+        elif isinstance(reason, diem_types.RefundReason__UserInitiatedFullRefund):
+            refund_reason = RefundReason.USER_INITIATED_FULL_REFUND
+        else:
+            refund_reason = RefundReason.OTHER
```

Comment:
add a class func to RefundReason for above logic, convert diem_types.RefundReason into DRW RefundReason